### PR TITLE
Issue time estimate, meaningful time tracking

### LINKED ISSUE #23112: Issue time estimate, meaningful time tracking
### Feature Description

1. Add functionality to set estimated time for issues similar to Jira (in "1w 3d 15h 30m" format)
2. Allow to view exact tracked time both logged, total and per-user (estimated time is not useful for tracking)
3. Replace Stopwatch "Add time" form from Hours and MInutes to Jira-like input in "1w 3d 15h 30m" format
4. Add ability to localize all time tracking issue comments in the future
5. Removes meaningless seconds from tracking display information
6. Makes time logging comments the same style as other comments (no new line and icon)

#### Why?
Time estimation is essential for some teams to know the importance of issues, the amount of effort an issue would take. And for business to know if it's profitable to spend money on developers time to implement an issue

More info on https://support.atlassian.com/jira-software-cloud/docs/estimate-an-issue/#Concepts-about-estimation

### Screenshots
image
image

### PR DESCRIPTION
**Pull request by stuzer05**
### This pr does:
1. Adds functionality to set estimated time for issues similar to Jira (in "1w 3d 15h 30m" format)
2. Now it's possible to view exact tracking time both logged, total and per-user (estimated time is not useful for tracking)
3. Replaces Stopwatch "Add time" form from Hours and MInutes to Jira-like input in "1w 3d 15h 30m" format
4. Adds ability to localize all time tracking issue comments in the future
5. Removes meaningless seconds from tracking display information
6. Makes time logging comments the same style as other comments (no new line and icon)

image
image

### Why?
Time estimation is essential for some teams to know the importance of issues, the amount of effort an issue would take. And for business to know if it's profitable to spend money on developers time to implement an issue

More info on https://support.atlassian.com/jira-software-cloud/docs/estimate-an-issue/#Concepts-about-estimation

Closes #23112

### CODE REVIEW

File: models/issues/comment.go
Comment (yardenshoham) [inline models/issues/comment.go]: I see no xorm annotation here, shouldn't these have it?

File: options/locale/locale_ru-RU.ini
Comment (yardenshoham) [inline options/locale/locale_ru-RU.ini]: Don't make changes to locale files other than en-US

File: options/locale/locale_uk-UA.ini
Comment (yardenshoham) [inline options/locale/locale_uk-UA.ini]: Don't make changes to locale files other than en-US

File: routers/web/repo/issue.go
Comment (yardenshoham) [inline routers/web/repo/issue.go]: > change issue's title

Copy-paste error?

File: services/issue/issue.go
Comment (yardenshoham) [inline services/issue/issue.go]: Copy-paste

File: services/issue/issue.go
Comment (yardenshoham) [inline services/issue/issue.go]:
```suggestion
```

File: services/issue/issue.go
Comment (yardenshoham) [inline services/issue/issue.go]:
```suggestion
```

**Review by yardenshoham — CHANGES_REQUESTED** [review]
Please make sure you've read https://github.com/go-gitea/gitea/blob/main/CONTRIBUTING.md

File: models/issues/comment.go
Comment (wxiaoguang) [reply models/issues/comment.go]: By default a field doesn't need the XORM tag if the field can be recognized as a column.

If this feature could be accepted on Gitea's roadmap, I have a question about this design, are these two columns (`hours+minutes`) easy to be used?

I guess one column `PlanTimSeconds` could be easier for calculation.

Comment (stuzer05) [reply models/issues/comment.go]: > If this feature could be accepted on Gitea's roadmap, I have a question about this design, are these two columns (`hours+minutes`) easy to be used?
> 
> I guess one column `PlanTimSeconds` could be easier for calculation.

Maybe you're right. Currently jira has only one field 
image

I can implement this behaviour and store the time is seconds. This should even better.

Comment (yardenshoham) [reply models/issues/comment.go]: "Estimate" is better than "plan time/time plan" IMO

Comment (stuzer05) [reply models/issues/comment.go]: > "Estimate" is better than "plan time/time plan" IMO

Agree

Comment (stuzer05) [reply models/issues/comment.go]: Hours and minutes fields were added to reflect the "Add time" pop-up functionality 
image

Comment (stuzer05) [reply models/issues/comment.go]: Maybe it's worth to replace whese fields with one, where user can enter jira-like time e.g. "3w 4d 12h". Both in estimate and "add time" interfaces

Comment (stuzer05) [reply models/issues/comment.go]: Currently, Jira doc says:
- You can specify a time unit after a time value 'X', such as Xw, Xd, Xh or Xm, to represent weeks (w), days (d), hours (h) and minutes (m), respectively.
- If you do not specify a time unit, hour will be assumed.
- Your current conversion rates are 1w = 5d and 1d = 8h.

**Comment by yardenshoham** [comment]
I suggest you remove the issue number from the title and add `Closes #23112` to the PR description. Also mark this PR as draft while you're working on it.

**Comment by stuzer05** [comment]
Ready functionality for the moment:
- Time estimate tracking
- Time tracker "Add time" input has been changed to match the one in Jira (e.g. "1h 50m")

Empty issue
image

Setting time estimate
image

Removing time estimate
image

Adding time to time tracker
image

image

**Comment by stuzer05** [comment]
I also think time tracker should display the real tie spent, not estimated. And the time in comments should be localized as well. Should I change that as well? @yardenshoham @wxiaoguang

**Comment by wxiaoguang** [comment]
> I also think time tracker should display the real time spent, not estimated. 

There is a `Time Tracker` feature (Enable Time Tracking on the repo setting page), does it do the same thing? 

<details>

image

</details>

> And the time in comments should be localized as well. Should I change that?

IIRC there is a HTML tag and JS for similar work. I am not sure whether it could help.

<details>

image

</details>

**Comment by stuzer05** [comment]
> IIRC there is a HTML tag and JS for similar work. I am not sure whether it could help.

image
All arrows pointing to time logged but not localized

The blue arrows pointing to the places where the real time should be displayed (the real value was "2w 15h" or smth)

**Comment by stuzer05** [comment]
> There is a `Time Tracker` feature (Enable Time Tracking on the repo setting page), does it do the same thing?

It's not always convenient to use a timer based time tracker but log time post-factum. It should be displayed fully in the comments

**Comment by wxiaoguang** [comment]
> > There is a `Time Tracker` feature (Enable Time Tracking on the repo setting page), does it do the same thing?
> 
> It's not always convenient to use a timer based time tracker but log time post-factum. It should be displayed fully in the comments

I am not a real user of such function, so I couldn't suggest more 😁  maybe you could try to write a detailed design document (in the PR main comment), then many people who have similar experiences could help.

**Comment by stuzer05** [comment]
The only thing I cannot figure out yet is how to make utils.SecToTime language-aware? How to call localization function outside of templates? @wxiaoguang

**Comment by wxiaoguang** [comment]
> The only thing I cannot figure out yet is how to make utils.SecToTime language-aware? How to call localization function outside of templates?

If I remember correctly, there is no such support yet.

**Comment by stuzer05** [comment]
> If I remember correctly, there is no such support yet.

Alright, I'll figure something out!

**Comment by yardenshoham** [comment]
There's a proposal https://github.com/tc39/proposal-intl-duration-format

**Comment by stuzer05** [comment]
> There's a proposal https://github.com/tc39/proposal-intl-duration-format

Awesome. Shoud I use it instead?

**Comment by yardenshoham** [comment]
It's not part of javascript yet

**Comment by stuzer05** [comment]
> It's not part of javascript yet

Then I'll use back-end localization for now. Gitea has all the necessaary translations already

**Comment by wxiaoguang** [comment]
Thanks for asking review from me, but my review doesn't count, and I am not a real user. Maybe maintainers could help.

**Comment by yardenshoham** [comment]
I don't have enough knowledge about the codebase to approve

**Comment by jolheiser** [comment]
I really think this should gather more feedback, either here or in the linked issue (like @wxiaoguang mentioned above with a design document)

As I said in the issue, I'm certainly willing to defer to the community, but I don't (personally) understand the value of this. I am obviously not a one-man roadblock, but I am curious to know what others think.

**Comment by stuzer05** [comment]
> I really think this should gather more feedback, either here or in the linked issue (like @wxiaoguang mentioned above with a design document)
> 
> As I said in the issue, I'm certainly willing to defer to the community, but I don't (personally) understand the value of this. I am obviously not a one-man roadblock, but I am curious to know what others think.
> 
> To clarify, I just want to make sure this is something that will get merged before you spend a lot of time on it. I am of course very glad you want to help with the project. slightly_smiling_face

Sure, I tried to explain why this feature is needed both in this pr and the issue. Besides, this pr also is a quality of life update for already existing time tracking.

File: models/issues/issue.go
Comment (yardenshoham) [inline models/issues/issue.go]: I think you need to add a database migration. From `CONTRIBUTING.md`:
> PRs that changes a database structure shall also have a migration step.

Comment (stuzer05) [reply models/issues/issue.go]: Thank you for pointing that out! Added migrations

**Comment by yardenshoham** [comment]
Also after this:
https://github.com/go-gitea/gitea/blob/f4920c9c7f5947d3b6476610f39bc3492ab4ef3b/models/migrations/migrations.go#L469

**Comment by stuzer05** [comment]
> Also after this:
> 
> https://github.com/go-gitea/gitea/blob/f4920c9c7f5947d3b6476610f39bc3492ab4ef3b/models/migrations/migrations.go#L469

Done!

**Comment by yardenshoham** [comment]
Your functions should actually be in a new file `v245.go`, don't modify old migrations

**Comment by stuzer05** [comment]
> Your functions should actually be in a new file `v245.go`, don't modify old migrations

Good point! Done

**Comment by wolfogre** [comment]
I agree with @jolheiser's comment on https://github.com/go-gitea/gitea/pull/23113#issuecomment-1444098901.

I don't fully understand the value of this feature either. To clarify, I don't believe it's entirely useless, but I am uncertain whether it's worth the effort. It's not just about one PR; there's also the matter of maintaining the feature and ensuring API/Cli support in the future. Reverting it would be nearly impossible once a new `CommentType` is introduced.

As @jolheiser said, this should gather more feedback.

**Comment by stuzer05** [comment]
> I agree with @jolheiser's comment on [#23113 (comment)](https://github.com/go-gitea/gitea/pull/23113#issuecomment-1444098901).
> 
> I don't fully understand the value of this feature either. To clarify, I don't believe it's entirely useless, but I am uncertain whether it's worth the effort. It's not just about one PR; there's also the matter of maintaining the feature and ensuring API/Cli support in the future. Reverting it would be nearly impossible once a new `CommentType` is introduced.
> 
> As @jolheiser said, this should gather more feedback.

Well, this feature is the only thing which is missing for time tracking, concidering there's another issue which adds statistics https://github.com/go-gitea/gitea/pull/19808 , this would greatly improve functionality of gitea in terms of replacement many other systems like Jira. Besides, time tracking is already an optional thing in gitea. Honestly, I don't see an issue to make time traking complete

I would love to maintain this feature. I use gitea for a long time already

File: models/issues/issue.go
Comment (yardenshoham) [inline models/issues/issue.go]: These can be defined globally before the function definition, similar to 
https://github.com/go-gitea/gitea/blob/cbc9a0fe477b1b8af249ca0b8dac5fc2be64e9f6/modules/templates/helper.go#L787
```suggestion

var (
	// Time match regex
	rOnlyHours = regexp.MustCompile(`^([\d]+)$`)
	rWeeks = regexp.MustCompile(`([\d]+)w`)
	rDays = regexp.MustCompile(`([\d]+)d`)
	rHours = regexp.MustCompile(`([\d]+)h`)
	rMinutes = regexp.MustCompile(`([\d]+)m`)
)

// TimeEstimateFromStr returns time estimate in seconds from formatted string
func (issue *Issue) TimeEstimateFromStr(timeStr string) int64 {
	timeTotal := 0
```

Comment (stuzer05) [reply models/issues/issue.go]: If these vars are used only once pre request and only in this function, why move them to the package scope?

Comment (delvh) [reply models/issues/issue.go]: Because you recalculate them with every request at the moment.
With the package scope, you only calculate them once.

Comment (stuzer05) [reply models/issues/issue.go]: > Because you recalculate them with every request at the moment. With the package scope, you only calculate them once.

Ah, alright. Thank you

File: routers/web/repo/issue.go
Comment (yardenshoham) [inline routers/web/repo/issue.go]: Move regex to package scope

Comment (stuzer05) [reply routers/web/repo/issue.go]: Done

File: models/issues/issue.go
Comment (lafriks) [inline models/issues/issue.go]: These functions should be moved to some util package. Also don't we have a similar function when PR is merged/commit pushed that also parses time spent in format ex `@1h`?

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/issues/issue.go
Comment (wxiaoguang) [reply models/issues/issue.go]: TBH, before asking for change, is there a conclusion that this feature could be merged or not?

If there is a consensus that this PR will be merged soon, then I can also help to review.

If there is no consensus, even if the requested changes are done, would a PR still get stale?

Comment (stuzer05) [reply models/issues/issue.go]: At least it's in 1.20 milestone. So I have hope it'll be merged. I have no clue where to discuss it. And as I sad in the issue itself, current time tracking is meaningless for any corporate use if there's no estimation

Comment (lafriks) [reply models/issues/issue.go]: I'm not against adding this feature, I would probably even use it :)

**Comment by lunny** [comment]
Maybe we need a better sidebar design, there are too items for time.

**Comment by stuzer05** [comment]
> Maybe we need a better sidebar design, there are too items for time.

I'd would love to work on new time tracking/estimate ui in the next pr, since I don't have ideas just yet and I'm really tired of merging conflicts from the main branch. Merging this one would be of huge help tbh 😅

**Comment by wxiaoguang** [comment]
@go-gitea/maintainers  especially @go-gitea/technical-oversight-committee 

Go or no-go for this PR? I know there were some discussions, but I think there is still no clear conclusion.

Since this PR is still active and well-maintained, from my perspective, if some nits could be fixed and some core code could be well tested (by test code), I think it's good to have.

But before spending more time on improving and writing more test code, could there be a clear conclusion? To save everyone's time.

File: public/assets/img/webpack/jquery.minicolors.0e614115.png
Comment (lunny) [inline public/assets/img/webpack/jquery.minicolors.0e614115.png:1]: I think this file should not be commited.

**Comment by lunny** [comment]
I would like to have this feature but I think something needs to be adjustment and discussed.
- What's the relation between `due date` and `estimate`? Looks like both them are for `planning`. Or maybe we could just have a `plan start date`?
- If we have estimate, then the issue siderbar will have `estimate`, `track time` and `due date`. I would like have a careful UI design to make it more clear, organized and use less places.

**Comment by lafriks** [comment]
While I would find this very useful I would like to get an actual concept on how we measure this. And how they correlate with a timer. In estimates week would usually mean working week example like 5 days * 8 hours (this should probably be server/org setting) but that would need to be taken into account when calculating and displaying actual time spent so that they do correlate correctly when looking at them side by side.

**Comment by techknowlogick** [comment]
Thanks for this PR!

 @lunny I don't think this is duplicating existing functionality.

* estimate = how much time I think this will take
* tracking = how much time this is taking
* due date = the date when this needs to/should be completed

I see that @lafriks has posted a comment, I'm going to go read that and then come back and update my comment with how I think we should proceed.

**Comment by yp05327** [comment]
Can we merge these three time tracking related things into one button in the sidebar, and then show a modal.
There are too many things in the sidebar.

**Comment by stuzer05** [comment]
> Thanks for this PR!
> 
> @lunny I don't think this is duplicating existing functionality.
> 
>     * estimate = how much time I think this will take
> 
>     * tracking = how much time this is taking
> 
>     * due date = the date when this needs to/should be completed
> 
> 
> I see that @lafriks has posted a comment, I'm going to go read that and then come back and update my comment with how I think we should proceed.

This! This is exactly what all this means

File: public/assets/img/webpack/jquery.minicolors.0e614115.png
Comment (silverwind) [reply public/assets/img/webpack/jquery.minicolors.0e614115.png:1]: Yes, remove it from the commit. It's in gitignore again since https://github.com/go-gitea/gitea/pull/30451.

**Comment by lunny** [comment]
I have the same idea. We need a design to include all three into the right sidebar but looks simpler. It took too much spaces currently.

**Review by wxiaoguang — COMMENTED** [review]
If I understand correctly: the feature is good, the UI needs some new design.

I will try to make some improvements later (especially the UI part, to reduce UI complexity), maybe we could try to get this in 1.23

**Comment by wxiaoguang** [comment]
@lafriks the "change request" seems outdated (https://github.com/go-gitea/gitea/pull/23113#pullrequestreview-1390997721), I think we could dismiss it.

**Comment by lafriks** [comment]
Not really, my concern was not addressed at all, currently estimate if I set a 5 days will be 24*5 hours that is not how estimates are expected in project management and this way they cannot be compared to time tracking as actual time spent to be useful. In project management when you estimate that the task requires 5 days to implement usually mean that 5 working days (5*8h or 5*7h etc depending on the company/country etc). My proposal would be to display estimate only in hours&minutes and later on implement other math behind this using org/global settings etc

**Comment by wxiaoguang** [comment]
> My proposal would be to display estimate only in hours&minutes

Agree, hour&minute is good enough for this feature IMO. Does anyone else have other suggestions? If no, I will do it this way.

**Comment by wxiaoguang** [comment]
> > My proposal would be to display estimate only in hours&minutes
> 
> Agree, hour&minute is good enough for this feature IMO. Does anyone else have other suggestions? If no, I will do it this way.

Update: sorry but I guess I wouldn't have enough time on this issue, recently there are many things in life that need to be dealt with so I can't promise when I would have time. Feel free to work on it if any one has time 🙏

File: models/issues/issue.go
Comment (6543) [inline models/issues/issue.go]: can you inject the context here

File: routers/web/repo/issue.go
Comment (6543) [inline routers/web/repo/issue.go]: could you use RenderedContent instead?

File: modules/util/time_str.go
Comment (6543) [inline modules/util/time_str.go]: adding tests for this function would be awesome!

File: templates/mail/issue/default.tmpl
Comment (6543) [inline templates/mail/issue/default.tmpl]: unrelated

File: models/issues/issue.go
Comment (6543) [reply models/issues/issue.go]: https://github.com/go-gitea/gitea/pull/23113/commits/823d67f27892bf744f257a098709745f38633e6d

File: modules/util/time_str.go
Comment (stuzer05) [reply modules/util/time_str.go]: Will work on it

**Comment by wxiaoguang** [comment]
I have some ideas to improves the UI now (after refactoring the issue sidebar)

Will work on it to address all the concerns above:

1. reduce UI complexity
2. simplify the time unit
3. add some tests

Then I think we can get it in 1.23

**Comment by wxiaoguang** [comment]
Almost done from my side. I think the new UI is better than before. 

image

image

image

image

**Comment by lafriks** [comment]
@wxiaoguang why still days? It still will be wrongly interpreted this way (1 astronomical day != 1 work day). Apart from that looks good code wise

**Comment by stuzer05** [comment]
> @wxiaoguang why still days? It still will be wrongly interpreted this way (1 astronomical day != 1 work day). Apart from that looks good code wise

work day could be different as well. Astronomical day is there to easily count hours

**Comment by stuzer05** [comment]
> I have some ideas to improve the UI now (after refactoring the issue sidebar)
> 
> Will work on it to address all the concerns above:
> 
> 1. reduce UI complexity (lunny: Maybe we need a better sidebar design, there are too items for time.)
> 2. simplify the time unit: (lafriks: My proposal would be to display estimate only in hours&minutes and later on implement other math behind this using org/global settings etc)
> 3. add some tests (6543: adding tests for this function would be awesome!)
> 
> Then I think we can get it in 1.23

Thank you for continuing on this!

**Comment by wxiaoguang** [comment]
> @wxiaoguang why still days? It still will be wrongly interpreted this way (1 astronomical day != 1 work day). Apart from that looks good code wise

Oops, I made a mistake. Please take a look at 9377c1da56110c74a9914145c541d3c11d3f211d , and I added enough comments there.

```
// When tracking working time, only hour/minute/second units are accurate and could be used.
// For other units like "day", it depends on "how many working hours in a day: 6 or 7 or 8?
// So at the moment, we only support hour/minute/second units.
// In the future, it could be some configurable options to help users
// to convert the working time to different units.
```

**Review by wxiaoguang — APPROVED** [review]

**Review by lunny — APPROVED** [review]
Unit test failure is related, otherwise LGTM

**Comment by wxiaoguang** [comment]
> Unit test failure is related, otherwise LGTM

CI passes, last call for reviews (especially the reviewers who have concerns and change-requests)

**Comment by lunny** [comment]
@lafriks needs your review

**Review by lafriks — APPROVED** [review]
Looks good, thanks 😉 

I don't see much point in seconds either as I don't see how any estimate could be less than 5 minutes or so exact but it's not blocking

**Comment by lunny** [comment]
I will merge this one manually if there is no response from @yardenshoham . I couldn't find a real unresolved block message.

**Review by yardenshoham — APPROVED** [review]

**Comment by wxiaoguang** [comment]
Need to follow up: The setting of the estimated time for time tracking in the issue should also have the same UI as adding time. #33318

**Pull request merged**
